tasks/Task.cpp

```diff
@@ -2,48 +2,496 @@
+            Json::Value message;
+            message["protocol"] = "one-to-one";
+            message["to"] = mRemotePeerID;
+            message["action"] = "candidate";
+            message["data"]["from"] = _local_peer_id.get();
+            message["data"]["candidate"] = string(candidate);
+            message["data"]["mid"] = candidate.mid();
+
+            if (auto ws = make_weak_ptr(mWs).lock())
+            {
+                Json::FastWriter fast;
+                ws->send(fast.write(message));
+            }
+        });
+
+    return peer_connection;
+}
+
+void Task::configurePeerDataChannel()
+{
+    mPeerConnection->onDataChannel(
+        [&](shared_ptr<rtc::DataChannel> data_channel)
+        {
+            mDataChannel = data_channel;
+            registerDataChannelCallBacks(data_channel);
+        });
+}
+
+void Task::configureWebSocket()
+{
+    std::promise<void> ws_promise;
+    auto ws_future = ws_promise.get_future();
+
+    mWs->onOpen(
+        [&]()
+        {
+            LOG_INFO_S << "WebSocket connected, signaling ready" << std::endl;
+            mState.web_socket = WebSocketOpened;
+            ws_promise.set_value();
+        });
+
+    mWs->onError(
+        [&](const string &error)
+        {
+            LOG_ERROR_S << "WebSocket failed: " << error << endl;
+            mState.web_socket = WebSocketFailed;
+            ws_promise.set_exception(std::make_exception_ptr(std::runtime_error(error)));
+            mPeerConnectionClosePromise.set_exception(std::make_exception_ptr(std::runtime_error(error)));
+            trigger();
+        });
+
+    mWs->onClosed(
+        [&]()
+        {
+            LOG_INFO_S << "WebSocket closed" << std::endl;
+            mState.web_socket = WebSocketClosed;
+            trigger();
+        });
+
+    mWs->onMessage(
+        [&](variant<binary, string> data)
+        {
+            if (!holds_alternative<string>(data))
+            {
+                return;
+            }
+
+            parseIncomingMessage(get<string>(data).c_str());
+            string actiontype = mDecoder.getActionType();
+
+            if (actiontype == "ping")
+            {
+                pong();
+            }
+            if (actiontype == "pong")
+            {
+                if (_remote_peer_id.get() == mDecoder.getFrom())
+                {
+                    mRemotePeerAnswerReceived = true;
+                    mWaitRemotePeerPromise.set_value();
+                }
+            }
+
+            if (actiontype == "offer")
+            {
+                onOffer();
+            }
+            else if (actiontype == "answer")
+            {
+                onAnswer();
+            }
+            else if (actiontype == "candidate")
+            {
+                onCandidate();
+            }
+        });
+
+    // wss://signalserverhost?user=yourname
+    const string url = _signaling_server_name.get() + "?user=" + _local_peer_id.get();
+    mWs->open(url);
+    ws_future.get();
```

Comment:
You should have a timeout for this as well.

Actually, you should have a timeout for *every* .get() you call on a future